rs/src/connections/host_relay.rs

```diff
@@ -0,0 +1,1085 @@
+                error: e,
+                reason: "could not unregister relay",
+            })
+    }
+
+    /// Adds a new port to the relay and returns a receiver for connections
+    /// that are made to that port. This is a "low level" type that you can use
+    /// if you want to deal with forwarding manually, but the `add_port` method
+    /// is appropriate and simpler for most use cases.
+    ///
+    /// Calling this method multiple times with the same port will result in
+    /// an error. Dropping the receiver **will not** remove the port, you must
+    /// call `remove_port()` to do that.
+    pub async fn add_port_raw(
+        &self,
+        port_to_add: &TunnelPort,
+    ) -> Result<mpsc::UnboundedReceiver<ForwardedPortConnection>, TunnelError> {
+        let n = port_to_add.port_number as u32;
+        if self.ports_tx.borrow().get(&n).is_some() {
+            return Err(TunnelError::PortAlreadyExists(n));
+        }
+
+        let tunnel_port = self
+            .mgmt
+            .create_tunnel_port(&self.locator, port_to_add, NO_REQUEST_OPTIONS)
+            .await;
+
+        match tunnel_port {
+            // created the port:
+            Ok(_) => {}
+            // the port's already registered, nothing we need to do:
+            Err(HttpError::ResponseError(e)) if e.status_code == 409 => {}
+            Err(e) => {
+                return Err(TunnelError::HttpError {
+                    error: e,
+                    reason: "failed to add port to tunnel",
+                })
+            }
+        }
+
+        let (tx, rx) = mpsc::unbounded_channel();
+        self.ports_tx.send_modify(|v| {
+            v.insert(n, tx);
+        });
+
+        Ok(rx)
+    }
+
+    /// Adds a new port to the tunnel and forwards TCP/IP connections made
+    /// over that port to the local machine. Calling this method multiple times
+    /// with the same port will result in an error.
+    pub async fn add_port(&self, port_to_add: &TunnelPort) -> Result<(), TunnelError> {
+        let rx = self.add_port_raw(port_to_add).await?;
+
+        tokio::spawn(forward_port_to_tcp(
+            SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), port_to_add.port_number).into(),
+            rx,
+        ));
+
+        Ok(())
+    }
+
+    /// Removes a port from the tunnel connection. Any channel returned from
+    /// `add_port_raw`, and any connections made within `add_port`, will close
+    /// shortly after this is called.
+    pub async fn remove_port(&self, port_number: u16) -> Result<(), TunnelError> {
+        self.mgmt
+            .delete_tunnel_port(&self.locator, port_number, NO_REQUEST_OPTIONS)
+            .await
+            .map_err(|e| TunnelError::HttpError {
+                error: e,
+                reason: "failed to remove port from tunnel",
+            })?;
+
+        self.ports_tx.send_modify(|v| {
+            v.remove(&(port_number as u32));
+        });
+
+        Ok(())
+    }
+
```

Comment:
> that causes the host to retrieve the updated port list from the service and add/remove forwarded ports accordingly

I don't think I would want this to be one-step like that in the Rust SDK. In the SDK, we give clients the option of handling port traffic directly rather than forwarding everything over TCP. We take advantage of this in the VS Code CLI. As a results, ports always get started or ended by the consumer telling the SDK how to handle the port. I think this isn't something that other SDKs support, which may impact how we think about the API.

From my understanding of your comment, in pseudo code adding a port in Rust would be something like:

```ts
mgmt.add_port(port)
relay.refresh_ports()
relay.tcp_forward_port(port) // or something like relay.get_port_stream
```

It seems like an `add_port`-esque API is more ergonomic, turning those three calls into one. The relay could also refresh automatically if the host asks for a port that it doesn't know about--though it's still another failure case in the SDK's API which isn't present with the current `add_port` method.

Regardless  of the above, if we support client requests for a port, we'd probably need some event/handler to be given to the RelayTunnelHost to figure out how to handle it.

Reply:
Other SDKs give _tunnel client_ applications the option to handle port connection streams directly, but currently assume _tunnel hosts_ are always forwarding TCP sockets. I agree that it is good to give hosting applications that flexibility also, as long as the more common case of forwarding TCP sockets is still simple enough.

> if we support client requests for a port, we'd probably need some event/handler to be given to the RelayTunnelHost to figure out how to handle it

Yes, we have scenarios in which something other than the host application is in control of which ports get forwarded. That was the reason for adding the refresh-ports API: to enable any authorized client to use the management API to add/remove ports. (Eventually I would like to eliminate the need for the explicit call to refresh-ports and replace it with some kind of message from the relay service that gets sent to the host automatically after any port changes via the management API.)

> It seems like an `add_port`-esque API is more ergonomic, turning those three calls into one.

My thinking was that it was better if the SDK did not offer multiple ways of doing the same thing, and easier for the host application developer to understand how it works if those calls are separate. But I'd be open to keeping `add_port`/`remove_port` as convenience methods if they document how they are equivalent of the other API calls.

> The relay could save a call by refreshing automatically if the host asks for a port that it doesn't know about

I assume you meant "if a _client_ asks for a port that that [the host] doesn't know about". I considered that as an alternative to the explicit refresh-ports message, but the problem is it doesn't work for _removed_ ports. If a client asks for a port that was removed, then the host won't know to refresh -- unless it refreshes before accepting every connection, but that would be slow. Also later there could be other properties on a port that affect how the host handles connections (for example protocol=udp).

Reply:
This issue doesn't need to block the PR. There will be more changes (across all the SDKs) anyway as we continue to refine and expand the APIs in this area. This is good for now.